packages/go_router/lib/src/builder.dart

```diff
@@ -47,164 +47,285 @@ class RouteBuilder {
+        buildErrorPage(context, e, matchList.location),
       ];
     }
+  }
 
-    // we should've set pages to something by now
-    assert(pages != null);
+  _RecursiveBuildResult _buildRecursive(
+    BuildContext context,
+    RouteMatchList matchList,
+    int startIndex,
+    VoidCallback pop,
+    bool routerNeglect, // TODO(johnpryan): Is this working? Needs a test...
+    Key navigatorKey,
+    Map<String, String> params,
+  ) {
+    final List<Page<dynamic>> pages = <Page<dynamic>>[];
+    final List<_OutOfScopePage> pagesForOutOfScopeNavigator =
+        <_OutOfScopePage>[];
+
+    for (int i = startIndex; i < matchList.matches.length; i++) {
+      final RouteMatch match = matchList.matches[i];
 
-    // pass either the match error or the build error along to the navigator
-    // builder, preferring the match error
-    if (matches.isError) {
-      error = matches.error;
+      if (match.error != null) {
+        throw RouteBuilderError('Match error found during build phase',
+            exception: match.error);
+      }
+
+      final RouteBase route = match.route;
+      final Map<String, String> newParams = <String, String>{
+        ...params,
+        ...match.decodedParams
+      };
+      if (route is GoRoute) {
+        final GoRouterState state = buildState(match, newParams);
+        final Page<dynamic> page = buildGoRoutePage(context, state, match);
+
+        // If this GoRoute is for a different Navigator, add it to the
+        // pagesForOutOfScopeNavigator list instead.
+        if (route.navigatorKey != null) {
+          pagesForOutOfScopeNavigator
+              .add(_OutOfScopePage(page, route.navigatorKey!));
+        } else {
+          pages.add(page);
+        }
+      } else if (route is ShellRoute) {
+        final GoRouterState state = buildState(match, newParams);
+        final _RecursiveBuildResult result = _buildRecursive(
+            context,
+            matchList,
+            i + 1,
+            pop,
+            routerNeglect,
+            // Use a default unique key for the Navigator if none is provided.
+            route.shellNavigatorKey,
+            newParams);
+        final Widget child = result.widget;
+
+        final Page<dynamic> page = buildPage(context, state,
+            callRouteBuilder(context, state, match, childWidget: child));
+
+        // If this ShellRoute is for a different Navigator, add it to the
+        // pagesForOutOfScopeNavigator list instead.
+        if (route.navigatorKey != null) {
+          pagesForOutOfScopeNavigator
+              .add(_OutOfScopePage(page, route.navigatorKey!));
+        } else {
+          pages.add(page);
+        }
+
+        // If any descendent GoRoutes have pages that are out of scope for this
+        // navigator, Add them to the list of routes that are out of scope. If
+        // they are in scope, add them to the list of pages for this navigator.
+        final List<_OutOfScopePage> pagesOutOfScopeForChildNavigator =
```

Comment:
This takes O(n^2), you can use Map<navigatorKey, List<Page>> to make it O(n)

Reply:
Fixed: https://github.com/flutter/packages/pull/2453/commits/87c16ad24a825d9cb68573a20fae1150c5786874